### PR TITLE
Extend langchain embedding API

### PR DESCRIPTION
**Pull request by yuwenzho**
## Type of Change

feature
API changed or not

## Description

Extend langchain embedding API for ITREX model

## How has this PR been tested?

CI

## Dependency Change?

no

### CODE REVIEW

**Comment by hshen14** [comment]
Do we have the example on how to use OptimizedEmbedding? how to support customized embedding?

File: intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py [hunk 1]
```diff
@@ -1,211 +1,212 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Wrapper for parsing the uploaded user file and then make document indexing."""
+
+
+import os, re
+from typing import Dict, List
+from langchain.vectorstores.chroma import Chroma
+from langchain.docstore.document import Document
+from langchain.embeddings import GooglePalmEmbeddings
```
Comment (kevinintel) [inline intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py:24]: move to embedding?

Comment (yuwenzho) [reply intel_extension_for_transformers/neural_chat/pipeline/plugins/retrieval/indexing/indexing.py:24]: `GooglePalmEmbeddings` is not for huggingface models.

**Review by hshen14 — APPROVED** [review]

File: intel_extension_for_transformers/langchain/embeddings/optimized_instructor_embedding.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            self._load_t5_model(model_name_or_path, config, cache_dir, **model_args)
+        elif isinstance(config, MT5Config):
+            self._load_mt5_model(model_name_or_path, config, cache_dir, **model_args)
+        else:
+            self.auto_model = OptimizedModel.from_pretrained(model_name_or_path, 
+                                                             config=config, 
+                                                             cache_dir=cache_dir, 
+                                                             **model_args)
+
+class OptimizedInstructior(InstructorEmbedding.INSTRUCTOR):
```
Comment (lvliang-intel) [inline intel_extension_for_transformers/langchain/embeddings/optimized_instructor_embedding.py:65]: OptimizedInstructior====>OptimizedInstructor

File: intel_extension_for_transformers/langchain/embeddings/optimized_instructor_embedding.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+from transformers import T5Config, MT5Config
+
+from .optimized_sentence_transformers import OptimzedTransformer
+
+sentence_transformers = LazyImport("sentence_transformers")
+InstructorEmbedding = LazyImport("InstructorEmbedding")
+
+logger = logging.getLogger(__name__)
+
+class OptimizedInstructiorTransformer(InstructorEmbedding.INSTRUCTOR_Transformer):
```
Comment (lvliang-intel) [inline intel_extension_for_transformers/langchain/embeddings/optimized_instructor_embedding.py:33]: OptimizedInstructiorTransformer=====>OptimizedInstructorTransformer

File: intel_extension_for_transformers/langchain/embeddings/optimized_instructor_embedding.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        # Load the modules of sentence transformer
+        modules_json_path = os.path.join(model_path, 'modules.json')
+        with open(modules_json_path) as fIn:
+            modules_config = json.load(fIn)
+
+        modules = OrderedDict()
+        for module_config in modules_config:
+            if module_config['idx']==0:
+                logger.info('load Optimized InstructiorTransformer')
+                module_class = OptimizedInstructiorTransformer
```
Comment (lvliang-intel) [inline intel_extension_for_transformers/langchain/embeddings/optimized_instructor_embedding.py:104]: OptimizedInstructiorTransformer====>OptimizedInstructorTransformer

**Comment by XuhuiRen** [comment]
I think these optimized embedding should be an available option for the user to choose instead of a static API?
It should add a new file to import the optimized embedding. The origin langchain embedding API should also be maintained. 
It would be better give a new name to the optimized embedding instead of using the same as the origin langchain embedding.

**Comment by yuwenzho** [comment]
> I think these optimized embedding should be an available option for the user to choose instead of a static API? It should add a new file to import the optimized embedding. The origin langchain embedding API should also be maintained. It would be better give a new name to the optimized embedding instead of using the same as the origin langchain embedding.

Our optimized embedding can accept both the models compatible with the origin langchain embedding API and the ITREX optimized models. Using the same API name ensures consistency in usage, and users would only need to modify the module's path.

**Review by XuhuiRen — APPROVED** [review]

**Pull request merged**